### PR TITLE
add a "tuple" type (array schema with list of item subschemas)

### PR DESCRIPTION
**Pull request by ericgj**
This implements a builder, decoder, encoder, validation, and tests for array schema `items` with an array of schemas (as listed in #6).  

This was called _tuple typing_ in an [earlier version of the spec](https://tools.ietf.org/html/draft-zyp-json-schema-03#section-5.5), so I thought it made sense to call it `tuple` / `tupleItems`, but see if you have another suggestion.   

Note I wasn't sure about the fuzzer unless we had tuple2Fuzzer, tuple3Fuzzer, etc. or something, so left it `fail` for now.

Thanks for the excellent library, it's super easy to understand and use!

### CODE REVIEW

File: src/JsonSchema/Validator.elm
Comment (jwoudenberg) [inline src/JsonSchema/Validator.elm]: 😻 Nice!

File: src/JsonSchema/Validator.elm
Comment (jwoudenberg) [inline src/JsonSchema/Validator.elm]: Ah, that answers my previous question about how `maxItems` and `minItems` are supposed to work!

File: tests/specs/ValidatorSpec.elm
Comment (jwoudenberg) [inline tests/specs/ValidatorSpec.elm]: 😻 Thanks for adding some tests!

**Review by jwoudenberg — APPROVED** [review]
This is great stuff, thanks for your contribution!

I have one suggesting for an approach to get the Fuzzer aspect of it to work. Let me know if you've an interest to try that out, otherwise I'm happy to merge this as is!

> Thanks for the excellent library, it's super easy to understand and use!

Thanks, that's really nice to hear!

File: src/JsonSchema/Fuzz.elm
Comment (ericgj) [reply src/JsonSchema/Fuzz.elm]: Very clear - thanks!  It practically writes itself.  I'll add that when I have a moment.

**Comment by ericgj** [comment]
I think `minItems` and `maxItems` are intended to be independent conditions and don't interact with `items` but I could be wrong... will research it a little more.

I also think I should add `additionalItems` since that particularly relates to tuple typing, I see now. Nice benefit of having separate Array and Tuple types that we can make it a compile error to add `additionalItems` to an Array schema in the builder.

**Comment by ericgj** [comment]
Added the tuple fuzzer, and `additionalItems`. But note the fuzzer is not complete, see note in commit message.

**Comment by ericgj** [comment]
PS. I think I can do a relatively clean fuzzer implementation taking into account `minItems` and `maxItems` too, there's just a couple corner cases like if the number of `items` schemas is less than `minItems` and no `additionalItems` schema is provided.

**Comment by jwoudenberg** [comment]
Hey @ericgj, this is looking great!

If I understand the spec correctly, when `items` is shorter than `minItems` and no `additionalItems` is provided, any value will be excepted for the extra array elements right? I remember being in a situation before where the fuzzer needed to generate 'any' JSON Value. I think at some point the nice thing to do would be to create a helper fuzzer that can really go crazy generating random JSON. For now though it seems fine to just set those extra array items to `null` or something, I think that would make the tests pass again, right? I'm happy to merge this when it's green.

Thanks for the effort, this is a super useful change!

**Comment by ericgj** [comment]
Fuzzer was quite a nasty thing to implement in the end, with 16 basic cases plus subcases for comparing `minItems` and `maxItems` to the length of the `items` schemas. Maybe someone can see a simpler way. But it is working, with test cases which _I think_ cover all the realistic cases (but definitely worth another set of eyes or two!). Two things:

- The fuzzer generates as long a tuple as it can given the `items`, if `maxItems` is specified. So for instance if you have 3 `items` schemas, and `maxItems` is 4, it will generate 3 elements that satisfy the 3 schemas (plus zero or one more element). It won't variably generate a 2-tuple for instance, even though that's technically valid.

- As mentioned in the commit notes, there can be impossible state specified if minItems > maxItems, etc. This kind of bleeds out into the Fuzzer. For these I just dump out an empty value, on the assumption they would never actually be specified. Not sure what else to do.

**Comment by jwoudenberg** [comment]
Sorry for the wait, I just came back from a holiday.

This is amazing work, thank you so much for the PR! Merging this!

**Comment by jwoudenberg** [comment]
Published as 3.0.0

**Comment by ericgj** [comment]
Cool, thanks!  

For future consideration re. the "minItems > maxItems" issue, something like [elm-bounded-number](http://package.elm-lang.org/packages/jschomay/elm-bounded-number/latest)  might work, though it would have to deal with both bounds being `Maybe`s, and decoding might be tricky. If you think it's worth it, I could open a new issue for discussion with that idea.

**Comment by jwoudenberg** [comment]
Another idea: we could use `Fuzz.invalid`: http://package.elm-lang.org/packages/elm-community/elm-test/4.2.0/Fuzz#invalid

In short: Fuzzers already support the concept of being 'invalid'. For example, the fuzzer `Fuzz.oneOf []` is invalid, because the `oneOf` cannot choose one fuzzer from an empty list. The `invalid` fuzzer mentioned above can be used by libraries to create their own Fuzzers that, under certain circumstances, might be invalid. We could use it to provide a nice error message if you attempt to create fuzzer for a schema where `minItems > maxItems`.

**Comment by ericgj** [comment]
Much better idea, I like it.

**Pull request merged**
